### PR TITLE
Add an invertible quote() function : funsor -> str

### PR DESCRIPTION
**Pull request by fritzo**
Based on: #242 

This adds an invertible `quote()` helper for funsors and things involving funsors:
```py
f = ...   # a Funsor
assert_close(f, eval(quote(f)))  # or something like this for lazy funsors
```
This indents in a flake8-compatible way so output can be pasted in tests, e.g:
```py
Independent(
 Independent(
  Contraction(
   ops.nullop,
   ops.add,
   frozenset(),
   (Tensor(
     torch.tensor([[-0.6077086925506592, -1.1546266078948975, -0.7021151781082153, -0.5303535461425781, -0.6365622282028198, -1.2423288822174072, -0.9941254258155823, -0.6287292242050171], [-0.6987162828445435, -1.0875964164733887, -0.7337473630905151, -0.4713417589664459, -0.6674002408981323, -1.2478348016738892, -0.8939017057418823, -0.5238542556762695]], dtype=torch.float32),  # noqa
     (('time__BOUND_4',
       bint(2),),
      ('_event_1__BOUND_2',
       bint(8),),),
     'real'),
    Gaussian(
     torch.tensor([[[-0.3536059558391571], [-0.21779225766658783], [0.2840439975261688], [0.4531521499156952], [-0.1220812276005745], [-0.05519985035061836], [0.10932210087776184], [0.6656699776649475]], [[-0.39107921719551086], [-0.20241987705230713], [0.2170514464378357], [0.4500560462474823], [0.27945515513420105], [-0.0490039587020874], [-0.06399798393249512], [0.846565842628479]]], dtype=torch.float32),  # noqa
     torch.tensor([[[[1.984686255455017]], [[0.6699360013008118]], [[1.6215802431106567]], [[2.372016668319702]], [[1.77385413646698]], [[0.526767373085022]], [[0.8722561597824097]], [[2.1879124641418457]]], [[[1.6996612548828125]], [[0.7535632252693176]], [[1.4946647882461548]], [[2.642792224884033]], [[1.7301604747772217]], [[0.5203893780708313]], [[1.055436372756958]], [[2.8370864391326904]]]], dtype=torch.float32),  # noqa
     (('time__BOUND_4',
       bint(2),),
      ('_event_1__BOUND_2',
       bint(8),),
      ('value__BOUND_1',
       reals(),),)),)),
  'gate_rate__BOUND_3',
  '_event_1__BOUND_2',
  'value__BOUND_1'),
 'gate_rate_t',
 'time__BOUND_4',
 'gate_rate__BOUND_3')
```
After this PR, `pretty()` is implemented on top of `quote()`.

Note: Ideally this method would be called `repr`, but since `repr` isn't invertible for e.g. `torch.Tensor`s (which abbreviate output) or functions (which print a pointer) we need to special-case such classes.

## Questions for reviewers

- What should I do about alpha renaming? Of course the global gensym state won't be serialized.
  One workaround is to `s/__BOUND/_b/` in pasted code?

## Tested
- [x] add unit test for invertibility
- [x] fix test/examples/test_bart.py using output from `quote()`

### CODE REVIEW

File: funsor/cnf.py [hunk 1]
```diff
@@ -9,22 +9,11 @@
...
 from funsor.interpreter import recursion_reinterpret
-from funsor.ops import AssociativeOp, DISTRIBUTIVE_OPS
-from funsor.terms import Align, Binary, Funsor, Number, Reduce, Subs, Unary, Variable, \
-    eager, normalize, to_funsor
+from funsor.ops import DISTRIBUTIVE_OPS, AssociativeOp, NullOp, nullop
+from funsor.terms import Align, Binary, Funsor, Number, Reduce, Subs, Unary, Variable, eager, normalize, to_funsor
 from funsor.torch import Tensor
 
 
-class NullOp(AssociativeOp):
```
Comment (fritzo) [inline funsor/cnf.py]: I've moved `NullOp` to funsor.ops for consistency

File: test/examples/test_bart.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+def unpack_gate_rate_0(gate_rate):
+    return unpack_gate_rate(gate_rate)[0]
+
+
+def unpack_gate_rate_1(gate_rate):
+    return unpack_gate_rate(gate_rate)[1]
+
+
+def test_bart():
```
Comment (fritzo) [inline test/examples/test_bart.py]: I expect tests like these to be short lived, but they are invaluable as @eb8680 and @jpchen and @martinjankowiak and I collaborate on experiments.

File: test/test_torch.py [hunk 3]
```diff
@@ -13,6 +13,17 @@
 from funsor.torch import REDUCE_OP_TO_TORCH, Einsum, Tensor, align_tensors, torch_stack, torch_tensordot
 
 
+@pytest.mark.parametrize('output_shape', [(), (2,), (3, 2)], ids=str)
+@pytest.mark.parametrize('inputs', [(), ('a',), ('a', 'b'), ('b', 'a', 'c')], ids=str)
+def test_quote(output_shape, inputs):
```
Comment (eb8680) [inline test/test_torch.py]: Can you add a test like this for `einsum` expressions?  Maybe just add the following to `test/test_cnf.py`:

```py
@pytest.mark.parametrize('equation,plates', EINSUM_EXAMPLES)
@pytest.mark.parametrize('backend', ['torch', 'pyro.ops.einsum.torch_log'])
@pytest.mark.parametrize('einsum_impl', [einsum, naive_plated_einsum])
def test_normalize_einsum(equation, plates, backend, einsum_impl):
    inputs, outputs, sizes, operands, funsor_operands = make_einsum_example(equation)

    with interpretation(reflect):
        expr = einsum_impl(equation, *funsor_operands, backend=backend, plates=plates)

    expected = reinterpret(expr)
    actual = eval(funsor.quote(expr))
    assert_close(actual, expected)
```

File: test/test_terms.py [hunk 4]
```diff
@@ -69,6 +77,32 @@ def test_cons_hash():
...
 
 
+def check_quote(x):
+    s = funsor.quote(x)
+    assert isinstance(s, str)
+    y = eval(s)
+    assert x is y
+
+
+@pytest.mark.parametrize('interp', [reflect, lazy, eager], ids=lambda i: i.__name__)
```
Comment (eb8680) [inline test/test_terms.py]: nit: add `normalize` to this list

**Comment by eb8680** [comment]
> What should I do about alpha renaming? ...  One workaround is to s/__BOUND/_b/ in pasted code?

This seems like a reasonable workaround

**Comment by fritzo** [comment]
Thanks for reviewing!
